keps/sig-apps/4368-support-managed-by-label-for-batch-jobs/README.md

```diff
@@ -0,0 +1,1044 @@
+- the Job controller reconciles jobs with custom `managed-by` label when the feature gate is disabled
+- verify the label is immutable, both when the job is suspended or unsuspended; when the feature is enabled
+- enablement / disablement of the feature after the Job (with custom `managed-by` label) is created
+
+##### Integration tests
+
+The following scenarios are covered:
+- the Job controller reconciles jobs with the `managed-by` label equal to `job-controller.k8s.io`
+- the Job controller reconciles jobs without the `managed-by` label
+- the Job controller does not reconcile a job with any other value of the `managed-by` label
+- the Job controller reconciles jobs with custom `managed-by` label when the feature gate is disabled
+
+During the implementation more scenarios might be covered.
+
+##### e2e tests
+
+The feature does not depend on kubelet, so the functionality can be fully
+covered with unit & integration tests.
+
+We propose a single e2e test for the following scenario:
+- the Job controller does not reconcile a job with any other value of the `managed-by` label
+
+### Graduation Criteria
+
+#### Beta
+
+- skip synchronization of jobs when the `managed-by` label does not exist, or equals `job-controller.k8s.io`
+- unit & integration tests
+- The feature flag enabled by default
+
+#### GA
+
+- Address reviews and bug reports from Beta users
+- e2e test
+- Re-evaluate the ideas of improving debuggability (like [extended `kubectl`](#debuggability), [dedicated condition](#condition-to-indicated-job-is-skipped), or [events](#event-indicating-the-job-is-skipped))
+- Re-evaluate the support for mutability of the label
+- Lock the feature gate
+
+#### Deprecation
+
+- Remove the feature-gate in GA+2.
+
+### Upgrade / Downgrade Strategy
+
+#### Upgrade
+
+An upgrade to a version which supports this feature does not require any
+additional configuration changes. This feature is opt-in at the Job-level, so
+to use it users need to add the `managed-by` label to their Jobs.
+
+#### Downgrade
+
+A downgrade to a version which does not support this feature does not require
+any additional configuration changes. All jobs, including these that
+specified a custom value for `managed-by`, will be handled in the default
+way by the Job controller.
+
+<!--
+If applicable, how will the component be upgraded and downgraded? Make sure
+this is in the test plan.
+
+Consider the following in developing an upgrade/downgrade strategy for this
+enhancement:
+- What changes (in invocations, configurations, API use, etc.) is an existing
+  cluster required to make on upgrade, in order to maintain previous behavior?
+- What changes (in invocations, configurations, API use, etc.) is an existing
+  cluster required to make on upgrade, in order to make use of the enhancement?
+-->
+
+### Version Skew Strategy
+
+N/A. This feature is limited to control plane. Also, This feature doesn't
```

Comment:
What about when you're half-way through upgrade and you're running mixed kcms, where some are still on v1.29 and others v1.30?

Reply:
Looking at https://kubernetes.io/releases/version-skew-policy/#kube-controller-manager-kube-scheduler-and-cloud-controller-manager I'm not sure we support running kube-controller-manager in HA mode. I assume only one of them is the leader performing the operations. When api-servers are in HA mode, and are skewed, then the old one can let the label to be mutated.

Reply:
I think the point is that the kcm in an older version might be the leader, in which case it wouldn't understand that it has to skip reconciling the Job. There isn't much we can do here, I think, other than recommending users not to use the feature during an ongoing upgrade.

Reply:
I see, added a paragraph along the lines in this section. PTAL.